File: 0139-word-break/0139-word-break.js
/**
 * @param {string} s
 * @param {string[]} wordDict
 * @return {boolean}
 */
var wordBreak = function(s, wordDict) {
    const cache = new Array(s.length+1).fill(false)
    cache[s.length] = true

    for(let i=s.length-1; i>=0; i--){
        for(const word of wordDict){
            if(i+word.length <= s.length // <= because i and the first char of word overlap
            && word == s.substring(i, i+word.length)){
                cache[i] = cache[i+word.length]
            }
            if(cache[i]){
                break;
            }
        }
    }

    return cache[0]
};

// cache is array s.length + 1 // +1 for the base case
// start from the end
// for the word in word,
// if start + word.length is less than s and dict matches substring start, word.length
// then cache[start] is cache[start+word.length]
// and break
// return cache[0] 

File: 0139-word-break/0139-word-break.py
class Solution:
    def wordBreak(self, s: str, wordDict: List[str]) -> bool:
        word_set = set(wordDict)
        word_lengths = set([len(w) for w in wordDict])
        n = len(s)

        dp = [False] * (n+1)
        dp[0] = True

        for i in range(1, n+1):
            for l in word_lengths:
                if l <= i and dp[i-l] and s[i-l:i] in word_set:
                    dp[i] = True
                    break
        
        return dp[n]